doc/developer-guides/coding_guidelines.rst

```diff
@@ -695,6 +724,294 @@ Compliant example::
+    
+    /* In `showcase.c`: */
+    #include "showcase.h"
+    
+    uint32_t func_showcase(uint32_t param)
+    {
+        return param;
+    }
+
+.. rst-class:: non-compliant-code
+
+   Non-compliant example::
+
+       /* In `showcase.h`: */
+       uint64_t func_showcase(uint64_t param);
+       
+       /* In `showcase.c`: */
+       #include "showcase.h"
+       
+       uint32_t func_showcase(uint32_t param)
+       {
+               return param;
+       }
+
+
+FN-12: Banned functions shall not be used
+=========================================
+
+The following cases shall be covered:
+
+a) The following functions shall not be used, calloc, malloc, realloc, and free.
+   Dynamic memory allocation shall be replaced with static memory allocation;
+b) The following functions shall be used with restrictions, va_arg, va_start,
+   and va_end. Variadic functions shall only be used for the functions similar to
+   printf.
+
+Compliant example::
+
+    uint32_t showcase_array[32];
+
+.. rst-class:: non-compliant-code
+
+   Non-compliant example::
+
+       uint32_t *showcase_ptr = (uint32_t *)malloc(32U * sizeof(uint32_t));
+
+
+FN-13: A function with declaration shall have the corresponding definition
+==========================================================================
+
+Compliant example::
+
+    /* In `showcase.h`: */
+    /* declaration */
+    uint32_t func_showcase(uint32_t param);
+    
+    /* In `showcase.c`: */
+    #include "showcase.h"
+    
+    /* definition */
+    uint32_t func_showcase(uint32_t param)
+    {
+        return param;
+    }
+
+.. rst-class:: non-compliant-code
+
+   Non-compliant example::
+
+       /* In `showcase.h`: */
+       /* declaration */
+       uint32_t func_showcase(uint32_t param);
+       
+       /* There is no definition of `func_showcase` anywhere in all source files */
+
+
+FN-14: All defined functions shall be used
+==========================================
+
+All defined functions shall be used, either called explicitly or indirectly via
+its address. Otherwise, the function shall be removed.
+
+Compliant example::
+
+    /* In `showcase.h`: */
+    uint32_t func_showcase(uint32_t param);
+    
+    /* In `showcase.c`: */
+    #include "showcase.h"
+    
+    uint32_t func_showcase(uint32_t param)
+    {
+        return param;
+    }
+    
+    /* In `main.c`: */
+    #include "showcase.h"
+    
+    void main(void)
+    {
+        uint32_t showcase;
+    
+        showcase = func_showcase(32U);
+    }
+
+.. rst-class:: non-compliant-code
+
+   Non-compliant example::
+
+       /* In `showcase.h`: */
+       uint32_t func_showcase(uint32_t param);
+       
+       /* In `showcase.c`: */
+       #include "showcase.h"
+       
+       /* There is no usage of `func_showcase` anywhere in all source files */
+       uint32_t func_showcase(uint32_t param)
+       {
+           return param;
+       }
+
+
+FN-15: Returning a structure from a function shall be used with restrictions
```

Comment:
This requirement is really about pointers and not structures.  A function shall not return a pointer to a local object, either directly or within a returned structure or array.